BoneReconstructionPlanner/BoneReconstructionPlanner.py

```diff
@@ -553,120 +597,161 @@ def process(self,fibulaLine,mandibularCurve,planeList,initialSpace,betweenSpace)
-      transformFidA.SetName("Mandible2Fibula Transform%d_A" % i)
-      slicer.mrmlScene.AddNode(transformFidA)
-      transformFidB = slicer.vtkMRMLLinearTransformNode()
-      transformFidB.SetName("Mandible2Fibula Transform%d_B" % i)
-      slicer.mrmlScene.AddNode(transformFidB)
+      #Start transformations
+      rotAxis2 = fibulaZ
+      rotAxis2 = rotAxis2/np.linalg.norm(rotAxis2)
+      angleRad2 = vtk.vtkMath.AngleBetweenVectors(rotatedMandibleAxisX, fibulaX)
+      angleDeg2 = vtk.vtkMath.DegreesFromRadians(angleRad2)
+
+      #this vector is created to check if rotAxis is okey or should be opposite
+      doublyRotatedMandibleAxisX = [0,0,0]
+      rotation2 = [angleRad2,rotAxis2[0],rotAxis2[1],rotAxis2[2]]
+      vtk.vtkMath.RotateVectorByWXYZ(rotatedMandibleAxisX,rotation2,doublyRotatedMandibleAxisX)
+
+      difference = np.linalg.norm(fibulaX-doublyRotatedMandibleAxisX)
+      if (difference>0.01):
+        rotAxis2 = [-rotAxis2[0],-rotAxis2[1],-rotAxis2[2]]
+        rotation2 = [angleRad2,rotAxis2[0],rotAxis2[1],rotAxis2[2]]
+        vtk.vtkMath.RotateVectorByWXYZ(rotatedMandibleAxisX,rotation2,doublyRotatedMandibleAxisX)
+      self.rotTransformParameters2.append([rotAxis2,angleDeg2])
+      
+
+      mandiblePlane0ToFibulaPlaneATransformNode = slicer.vtkMRMLLinearTransformNode()
+      mandiblePlane0ToFibulaPlaneATransformNode.SetName("Mandible2Fibula Transform%d_A" % i)
+      slicer.mrmlScene.AddNode(mandiblePlane0ToFibulaPlaneATransformNode)
+      mandiblePlane1ToFibulaPlaneBTransformNode = slicer.vtkMRMLLinearTransformNode()
+      mandiblePlane1ToFibulaPlaneBTransformNode.SetName("Mandible2Fibula Transform%d_B" % i)
+      slicer.mrmlScene.AddNode(mandiblePlane1ToFibulaPlaneBTransformNode)
 
       if i==0:
-        planesPositionA.append(lineDirectionVersor*initialSpace)
+        fibula2FibulaPlanesPositionA.append(fibulaZ*initialSpace)
       else:
-        planesPositionA.append(planesPositionB[i-1] + lineDirectionVersor*betweenSpace)
+        fibula2FibulaPlanesPositionA.append(fibula2FibulaPlanesPositionB[i-1] + fibulaZ*betweenSpace)
       
-      planesPositionB.append(planesPositionA[i] + d[i]*lineDirectionVersor)
+      fibula2FibulaPlanesPositionB.append(fibula2FibulaPlanesPositionA[i] + boneSegmentsDistance[i]*fibulaZ)
 
-      finalTransformA = vtk.vtkTransform()
-      finalTransformA.PostMultiply()
-      finalTransformA.Translate(-lineStartPos[0], -lineStartPos[1], -lineStartPos[2])
-      finalTransformA.RotateWXYZ(angleDeg,rotAxis)
-      finalTransformA.Translate(lineStartPos)
-      finalTransformA.Translate(planesPositionA[i])
+      mandiblePlane0ToFibulaPlaneATransform = vtk.vtkTransform()
+      mandiblePlane0ToFibulaPlaneATransform.PostMultiply()
+      mandiblePlane0ToFibulaPlaneATransform.Translate(-mandiblePlane0Origin[0], -mandiblePlane0Origin[1], -mandiblePlane0Origin[2])
+      mandiblePlane0ToFibulaPlaneATransform.RotateWXYZ(angleDeg,rotAxis)

```

Comment:
don't use RotateWXYZ

Reply:
Are you sure we need to change RotateWXYZ? I use an euler axis and and euler angle to make each rotation:
https://en.wikipedia.org/wiki/Euler%27s_rotation_theorem
https://en.wikipedia.org/wiki/Rotation_formalisms_in_three_dimensions#Rotation_matrix_%E2%86%94_Euler_axis/angle

Are you very keen on using Euler angles (z-y′-x″ intrinsic) for rotations?
https://en.wikipedia.org/wiki/Rotation_formalisms_in_three_dimensions#Euler_angles_(z-y%E2%80%B2-x%E2%80%B3_intrinsic)_%E2%86%92_rotation_matrix

If so, can you send me a pdf of how to make the transforms I defined using Euler angles (z-y′-x″ intrinsic)?

Reply:
In medical image computing we generally avoid Euler angles but instead we always use transformation matrices. Euler angles are only usable in practice if you only ever rotate along 1 or 2 axes and the rotation is not more than 10-20 degrees along each axis. Otherwise you run into complications (ambiguous conversion between rotation angles and transformation matrix, gimbal lock, etc.). You can still use it for example to slightly adjust an orientation using sliders.

Reply:
Can you send me an example of a rotation matrix from X1Y1Z1 to X2Y2Z2?
For example, let suppose the rotation matrix from X1Y1Z1 to X2Y2Z2 is:
RotationMatrix = RotateZ(angleZ) * RotateY(angleY) * RotateX(angleX) -- (these are matrixes on a matrixes product)
How angleZ, angleY and angleX are defined having into account that you have the definition of X1Y1Z1 and X2Y2Z2?

Reply:
You can use vtkTransform's RotateX/RotateY/RotateZ, but preferably only two of them, and in a small range, and only for adjusting transforms using sliders on the GUI. For everything else, it is better to directly compute the transformation matrix axes (using cross/dot products, etc.).

Reply:
Oh, I think I understand now. You want me to calculate the projections of X2Y2Z2 over X1Y1Z1 to define the rotation matrix.

Reply:
Yes, any clear, explicit, robust way of computing the transformation matrix works (that can be easily checked against specification in Design.pptx).